src/expression/evaluation_context.ts

```diff
@@ -29,8 +40,47 @@ class EvaluationContext {
         return this.feature && 'id' in this.feature ? this.feature.id : null;
     }
 
+    geometryDollarType() {
+        return this.feature ?
+            typeof this.feature.type === 'number' ? geometryTypes[this.feature.type] : simpleGeometryType[this.feature.type] :
+            null;
+    }
+
     geometryType() {
-        return this.feature ? typeof this.feature.type === 'number' ? geometryTypes[this.feature.type] : this.feature.type : null;
+        let geometryType = this.feature.type;
+        if (typeof geometryType === 'number') {
+            geometryType = geometryTypes[this.feature.type];
+            if (geometryType !== 'Unknown') {
+                const geom = this.geometry();
+                const len = geom.length;
+                if (len > 1) {
+                    switch (geometryType) {
+                        case 'Point':
+                            geometryType = 'MultiPoint';
+                            break;
+                        case 'LineString':
+                            geometryType = 'MultiLineString';
+                            break;
+                        case 'Polygon':
+                            // Following https://github.com/mapbox/vector-tile-js/blob/77851380b63b07fd0af3d5a3f144cc86fb39fdd1/lib/vectortilefeature.js#L197
+                            for (let i = 0, ccw; i < len; i++) {
+                                const area = calculateSignedArea(geom[i]);
+                                if (area === 0) continue;
+                                if (ccw === undefined) {
+                                    ccw = area < 0;
+                                } else if (ccw === area < 0) {
+                                    geometryType = 'MultiPolygon';
+                                    break;
+                                }
+                            }
+                            break;
+                        default:
+                            break;
+                    }
+                }
+            }
+        }
+        return geometryType;
```

Comment:
Should this response be cached in order to avoid recalculating this? Or is it OK to calculate this each time as it's not really possible to cache anything?

Reply:
I'm not aware of such mechanism, besides maybe storing it in the geometry itself...? IDK...

Reply:
I'd rather not introduce a new optimization mechanism for a problem we don't know that has a significant impact.